xborders

```diff
@@ -199,85 +178,120 @@ class Highlight(Gtk.Window):
+    def _composited_changed_event(self, arg):
+        if self.screen.is_composited():
+            self.move(0, 0)
         else:
-            if (self.notified_compositor == True):
-                self.move(0, 0);
-            self.notified_compositor = False;
-        ctx.save()
-
-        if (not is_fullscreen(self.wnck_screen)):
-            win_data = get_win(self.wnck_screen)
-            if win_data != None:
-                x, y, w, h = win_data[0][0], win_data[0][1], win_data[1][0], win_data[1][1]
-                if BORDER_MODE == 0:
-                    x += BORDER_WIDTH / 2;
-                    y += BORDER_WIDTH / 2;
-                    w -= BORDER_WIDTH;
-                    h -= BORDER_WIDTH;
-                elif BORDER_MODE == 1:
-                    x -= BORDER_WIDTH / 2;
-                    y -= BORDER_WIDTH / 2;
-                    w += BORDER_WIDTH;
-                    h += BORDER_WIDTH;
-                draw.draw_rectangle(
-                    ctx,
-                    x,
-                    y,
-                    w,
-                    h,
-                    BORDER_RADIUS,
-                    BORDER_WIDTH,
-                    [0, 0, 0, 0],
-                    [BORDER_R, BORDER_G, BORDER_B, BORDER_A],
-                )
+            self.move(1e6, 1e6)
+            subprocess.Popen(["notify-send", "xborder", "xborders requires a compositor. Resuming once a compositor is running."])
+
+    # This event will trigger every active window change, it will queue a
+    # border to be drawn and then do nothing.
+    # See:
+    #    Signals available for the Wnck.Screen class: https://lazka.github.io/pgi-docs/Wnck-3.0/classes/Screen.html#signals
+    #    Signals available for the Wnck.Window class: https://lazka.github.io/pgi-docs/Wnck-3.0/classes/Window.html#signals
+    def _active_window_changed_event(self, screen, previous_active_window):
+        active_window = self.wnck_screen.get_active_window()
+
+        self.border_path = [0, 0, 0, 0]
+        if active_window is not None:
+            # Find if the window has a 'geometry-changed' event connected.
+            signal_id = GObject.signal_lookup('geometry-changed', active_window)
+            has_event_connected = GObject.signal_has_handler_pending(active_window, signal_id, 0, False)
+            
+            # if it doesn't have one.
+            if not has_event_connected:
+                # Connect it.
+                # Has to be done this way in order to not connect an event
+                # every time the active window changes, thus, drawing unnecesary frames.
+                active_window.connect('geometry-changed', self._geometry_changed_event)
+
+            if active_window.get_state() != Wnck.WindowState.FULLSCREEN:
+                self._calc_border_geometry(active_window)
+        self.queue_draw()
 
-        ctx.restore()
+    # This is weird, "window_changed" is not necessarily the active window,
```

Comment:
But this could be a good thing for us if we would like to implement the feature #16 
To implement that you only need to remove the if branch and use window_changed parameter instead of the active_window, right?

Reply:
Yes, window_changed is an Wnck.Window object, you can manage it like the active_window object and make out Gtk.Widget draw whatever you want into it.